### ISSUE TITLE
OnDuplicateKeyIgnore & OnDuplicateKeyIgnore 

### ISSUE THREAD

**Issue by name256** [xuejmnet/easy-query#132]
Good day,

I have been going through your product and was wondering whether you have the following functions for the different sql dialects. I believe PgSQL uses OnConflict

**Comment by xuejmnet** [comment]
Thank you for your interest in my product.

Sure, it is possible. The easy-query framework has already abstracted the expressions, and the subsequent step is to implement the SQL syntax for each specific database. However, for specific methods like insertOrUpdate, it is desirable to abstract them into the expression itself. Abstracting special SQL dialects into the expression might cause compatibility issues with certain databases that do not support them.

Therefore, the functionality you mentioned can be fully implemented. Currently, I have implemented the default support for MySQL syntax, and since PGSQL is compatible with MySQL syntax, it does not require any modifications to achieve most SQL operations. In the future, I will gradually implement support for other databases. Moreover, I welcome collaboration and discussions through pull requests (PRs) on how to abstract expressions for different databases. For example, easyQuery.insertable(A.class).usePgSQL().onConflict could be a possible approach.

<img width="1110" alt="image" src="https://github.com/xuejmnet/easy-query/assets/18061232/913089b1-257b-4bed-9d88-de1134abe94c">

**Comment by name256** [comment]
I will take some time studying the codebase and will let you know when I
can be available to assist on adding more features to the project. I
currently use a combination of products that sometimes make my code really
bulky and I wanted to have one product that could cater for what I require
since a lot of the projects are usually CRUD and not too complicated.

Kind regards
------
Abel Salama Birya
254727452403

**Comment by xuejmnet** [comment]
It would be my honor to help anyone with my product, and I will also be working on developing documentation in English for the future. Currently, I am focusing on improving the Chinese documentation.  ： ）

**Comment by name256** [comment]
Good day to you,

I have taken some time to look at the code as well as I can. Kindly let me
know how best I can be able to approach adding the 2 functions for MySQL &
PostgreSQL.

Kind regards
------
Abel Salama Birya
254727452403

**Comment by xuejmnet** [comment]
@name256 
<img width="1620" alt="image" src="https://github.com/xuejmnet/easy-query/assets/18061232/750a5a16-f887-49b5-99b0-f02589b283c9">

## first
```java
public interface SQLOnDuplicateKeyIgnore<TChain> {
    /**
     * 部分数据库不支持
     * @return
     */
    TChain onDuplicateKeyIgnore();
}
```
<img width="1222" alt="image" src="https://github.com/xuejmnet/easy-query/assets/18061232/cb23e080-fce2-47b4-8f6b-f082cc5aa56f">

## second
`Insertable` implments SQLOnDuplicateKeyIgnore
<img width="1662" alt="image" src="https://github.com/xuejmnet/easy-query/assets/18061232/82bf7842-ddfe-4a23-a5af-c6c00b833b32">


## third
<img width="1595" alt="image" src="https://github.com/xuejmnet/easy-query/assets/18061232/8d4bd332-e542-43ad-b680-f441b4be8c74">


## end
<img width="1503" alt="image" src="https://github.com/xuejmnet/easy-query/assets/18061232/be7fa9b1-75a3-4e8b-aff6-c543b9ea2bd1">

**Comment by xuejmnet** [comment]
@name256 thanks very much. I have already implemented these two functions for you and made them differentiate between the usage of API and SQL in expressions by the users.

**Comment by xuejmnet** [comment]
queuryable | insertable | updatable | deletable =====> expression builder ======> expression=====>sql

**Comment by name256** [comment]
Thank you for the update

I will update the code and let you know how it goes.

Kind regards
------
Abel Salama Birya
254727452403

**Comment by name256** [comment]
Good day,

I have tried using the library in my project but the latest version on
maven is 0.9.32. Were you able to commit the updated library to maven
central ro enable access to the onDuplicateKey methods?

Knd regards
------
Abel Salama Birya
254727452403

**Comment by xuejmnet** [comment]
@name256  https://central.sonatype.com/search?q=easy-query use 1.1.17

**Comment by name256** [comment]
Good day,

I have added the dependency to my project but I only seem to have the
onDuplicateKeyIgnore() and onConflictDoNothing() methods.

Kind regards
------
Abel Salama Birya
254727452403

**Comment by name256** [comment]
Kindly disregard the previous screenshot sent. Please find the correct one
attached.
------
Abel Salama Birya
254727452403

**Comment by xuejmnet** [comment]
[onConflictDoUpdate、onDuplicateKeyUpdate for mysql pgsql](https://github.com/xuejmnet/easy-query/commit/9a1c151437efb3904d2e5537927a6076a93aaa4f) this commit will publish next version (1.1.18) in this night

**Comment by xuejmnet** [comment]
@name256 I have already released version 1.1.18. Please check the corresponding website in 30 minutes to an hour to see if you can find the corresponding version. If so, Maven can be used to reference it. https://central.sonatype.com/search?q=easy-query

**Comment by name256** [comment]
I was able to download the newer version. Thank you.

However I believe that we may need to make updates on the PgSQL method
onConflictUpdate() due to allow the specification of the duplicate key
field.

INSERT INTO "books" ("name","created") VALUES ('Making Money',1688813216) *ON
CONFLICT ("id")* DO UPDATE SET "name" = EXCLUDED."name", "created" =
EXCLUDED."created" *was aborted: ERROR: duplicate key value violates unique
constraint "books_name"*

Therefore on the above insert I would specify *ON CONFLICT("name")* to
avoid the error above.

Kind regards
------
Abel Salama Birya
254727452403

**Comment by xuejmnet** [comment]
@name256 that's good idea next version i will support it

**Comment by xuejmnet** [comment]
@name256 I have already released version 1.1.19. Please check the corresponding website in 30 minutes to an hour to see if you can find the corresponding version. If so, Maven can be used to reference it. https://central.sonatype.com/search?q=easy-query

**Comment by xuejmnet** [comment]
@name256 unit test has samples
```java
        TopicAuto topicAuto = new TopicAuto();
        topicAuto.setStars(999);
        topicAuto.setTitle("title" + 999);
        topicAuto.setCreateTime(LocalDateTime.now().plusDays(99));
        Assert.assertNull(topicAuto.getId());
        EntityInsertable<TopicAuto> insertable = easyQuery.insertable(topicAuto).onConflictDoUpdate(TopicAuto::getTitle,t->t.column(TopicAuto::getStars).column(TopicAuto::getCreateTime));
        String sql = insertable.toSQL(topicAuto);
        Assert.assertEquals("INSERT INTO \"t_topic_auto\" (\"stars\",\"title\",\"create_time\") VALUES (?,?,?) ON CONFLICT (\"title\") DO UPDATE SET \"stars\" = EXCLUDED.\"stars\", \"create_time\" = EXCLUDED.\"create_time\"",sql);
    
```
